### PR TITLE
New lint: manual-range-contains

### LINKED ISSUE #1110: Lint wrong integer range comparisons
I don't know if an issue already exists for something like this, but at least I couldn't find one:
I suggest to add a lint that checks for conditions such as
- `i > A || i < B` where `A` and `B` are integer literals (or constants?) with `A` ≤ `B` (always true)
- `i < A && i > B` where `A` and `B` are integer literals (or constants?) with `A` ≤ `B` (always false)

After https://github.com/rust-lang/rust/issues/32311 is stabilized, clippy could also suggest to use `Range::contains`.

### PR DESCRIPTION
**Pull request by llogiq**
This fixes #1110, at least for the contains-suggesting part.

- \[x] Followed [lint naming conventions][lint_naming]
- \[x] Added passing UI tests (including committed `.stderr` file)
- \[x] `cargo test` passes locally
- \[x] Executed `cargo dev update_lints`
- \[x] Added lint documentation
- \[x] Run `cargo dev fmt`
---

changelog: new lint: manual-range-contains

### CODE REVIEW

**Comment by llogiq** [comment]
error seems unrelated, we may need another rustup.

**Comment by josephlr** [comment]
@llogiq I looks like we need to sync with rustc, and the process currently takes over and hour even after you patch git to make things work better (see https://github.com/rust-lang/rust-clippy/issues/5565). I'm generating the PR right now. I'll ping this issue when the PR is up.

**Comment by josephlr** [comment]
https://github.com/rust-lang/rust-clippy/pull/6178 is up.

**Comment by llogiq** [comment]
@josephlr So what do I need to do now? I've updated rust and pulled from master but still get an error.

**Comment by ebroto** [comment]
@llogiq it should work now if you rebase on master (#6184)

**Comment by llogiq** [comment]
rebased. Thanks @ebroto.

File: clippy_lints/src/ranges.rs [hunk 1]
```diff
@@ -173,6 +185,157 @@ impl<'tcx> LateLintPass<'tcx> for Ranges {
...
+            // `.iter()` and `.len()` called on same `Path`
+            if let ExprKind::Path(QPath::Resolved(_, ref iter_path)) = iter_args[0].kind;
+            if let ExprKind::Path(QPath::Resolved(_, ref len_path)) = len_args[0].kind;
+            if SpanlessEq::new(cx).eq_path_segments(&iter_path.segments, &len_path.segments);
+            then {
+                span_lint(cx,
+                RANGE_ZIP_WITH_LEN,
+                span,
+                &format!("it is more idiomatic to use `{}.iter().enumerate()`",
+                    snippet(cx, iter_args[0].span, "_")));
```
Comment (flip1995) [inline clippy_lints/src/ranges.rs:333]:
```suggestion
                span_lint(
                    cx,
                    RANGE_ZIP_WITH_LEN,
                    span,
                    &format!("it is more idiomatic to use `{}.iter().enumerate()`", snippet(cx, iter_args[0].span, "_"))
                );
```

File: tests/ui/range.rs [hunk 2]
```diff
@@ -14,3 +14,40 @@ fn no_panic_with_fake_range_types() {
 
     let _ = Range { foo: 0 };
 }
+
+#[warn(clippy::manual_range_contains)]
```
Comment (flip1995) [inline tests/ui/range.rs:18]: Can you move those to a new test file and add `// run-rustfix` there please?

File: clippy_lints/src/ranges.rs [hunk 3]
```diff
@@ -173,6 +185,157 @@ impl<'tcx> LateLintPass<'tcx> for Ranges {
...
+            };
+            if l_inc {
+                return;
+            }
+            let (range_type, range_op) = if u_inc {
+                ("Range", "..")
+            } else {
+                ("RangeInclusive", "..=")
+            };
+            span_lint_and_then(
```
Comment (flip1995) [inline clippy_lints/src/ranges.rs:254]: You can use `span_lint_and_sugg` here. Same above.

**Review by flip1995 — COMMENTED** [review]
Impl LGTM.

Also, since you will read this comment: You have some stale branches in the `rust-lang/rust-clippy` repository. We try to delete all branches, that aren't necessary for the infra: https://github.com/rust-lang/rust-clippy/issues/4745 Could you please move them to a fork and then delete them here in this repo?

**Comment by llogiq** [comment]
I don't think I'll need them anymore. Will delete them once I get to my PC.

File: tests/ui/range.rs (hunk 2 above)
Comment (llogiq) [reply tests/ui/range.rs]: Sure thing!

File: clippy_lints/src/ranges.rs (hunk 3 above)
Comment (llogiq) [reply clippy_lints/src/ranges.rs]: Done, thanks!

File: clippy_lints/src/ranges.rs (hunk 1 above)
Comment (llogiq) [reply clippy_lints/src/ranges.rs]: Somehow cargo dev fmt did this. Fixed.

**Comment by llogiq** [comment]
@flip1995 Done....and...we need yet another rustup.

**Comment by ebroto** [comment]
Should be fixed by #6189

**Comment by llogiq** [comment]
Cool. Now there seems to be yet another unrelated error!?

**Comment by llogiq** [comment]
Nevermind, I fixed it en passant. Seems to relate to `String::new()` now being const fn.

**Comment by llogiq** [comment]
@flip1995 r?

**Comment by llogiq** [comment]
Huh? There was some strange interaction (presumably with const stability) in the `or_fun_call` test. It's fixed now. r? @flip1995

**Comment by llogiq** [comment]
Since @flip1995 is currently busy, r? @ebroto

**Comment by flip1995** [comment]
@bors r+

Thanks!

**Pull request merged**